### PR TITLE
statistics: compute and store column order correlation with handle

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

First step to fix https://github.com/pingcap/tidb/issues/9067

### What is changed and how it works?

- compute order correlations between columns and handle column;
- store this correlation in mysql.stats_histograms, i.e, in a newly added column `Correlation`;
- update the upgrade logic and `show stats_histograms` correspondingly;

### Check List 

Tests 

 - Unit test

Code changes

 - Has persistent data change

Side effects

 - Increased code complexity
 - Breaking backward compatibility: upgrade logic has been addressed in this PR also.

Related changes

N/A

### CODE REVIEW

File: statistics/handle_test.go [hunk 1]
```diff
@@ -437,3 +437,52 @@ func newStoreWithBootstrap(statsLease time.Duration) (kv.Storage, *domain.Domain
 	do.SetStatsUpdating(true)
 	return store, do, errors.Trace(err)
 }
+
+func (s *testStatsSuite) TestCorrelation(c *C) {
+	defer cleanEnv(c, s.store, s.do)
+	testKit := testkit.NewTestKit(c, s.store)
+	testKit.MustExec("use test")
+	testKit.MustExec("create table t(c1 int primary key, c2 int)")
```
Comment (lamxTyler) [inline statistics/handle_test.go]: When there is a column that is the only part of one index, like `create table t(a int, b int, index idx(a))`, we will not create statistics for the column(Refer https://github.com/pingcap/tidb/blob/master/planner/core/planbuilder.go#L675). Maybe we need to remove that logic.

File: statistics/sample.go [hunk 2]
```diff
@@ -29,9 +30,58 @@ import (
...
+// SortSampleItems sorts a slice of SampleItem.
+func SortSampleItems(sc *stmtctx.StatementContext, items []*SampleItem) error {
+	sorter := sampleItemSorter{items: items, sc: sc}
+	sort.Sort(&sorter)
+	return sorter.err
+}
+
+// extractSampleItemsDatums is for test purpose only to extract Datum slice
+// from SampleItem slice.
+func extractSampleItemsDatums(items []*SampleItem) []types.Datum {
```
Comment (qw4990) [inline statistics/sample.go]: It's better to put this function in `fmsketch_test.go`?

Comment (eurekaka) [reply statistics/sample.go]: Updated, PTAL

File: statistics/builder.go [hunk 3]
```diff
@@ -143,14 +145,32 @@ func BuildColumn(ctx sessionctx.Context, numBuckets, id int64, collector *Sample
...
+	// we can simply compute sum(X) = sum(Y) =
+	//    (sampleNum-1)*sampleNum / 2
+	// and sum(X^2) = sum(Y^2) =
+	//    (sampleNum-1)*sampleNum*(2*sampleNum-1) / 6
+	// The formula for computing correlation is borrowed from PostgreSQL.
+	// Note that (itemsCount*corrX2Sum - corrXSum*corrXSum) would never be zero when sampleNum is larger than 1.
+	itemsCount := float64(sampleNum)
+	corrXSum := (itemsCount - 1) * itemsCount / 2.0
+	corrX2Sum := (itemsCount - 1) * itemsCount * (2*itemsCount - 1) / 6.0
+	hg.Correlation = (itemsCount*corrXYSum - corrXSum*corrXSum) / (itemsCount*corrX2Sum - corrXSum*corrXSum)
```
Comment (zz-jason) [inline statistics/builder.go]: How about pointing out this correlation is Pearson correlation coefficient, is calculated based on the formula specifies on https://en.wikipedia.org/wiki/Pearson_correlation_coefficient, which also used by PostgreSQL?

Comment (eurekaka) [reply statistics/builder.go]: Comments updated, PTAL

**Comment by eurekaka** [comment]
Wait for https://github.com/pingcap/tidb-test/pull/740 to be merged together.

**Pull request merged**